### PR TITLE
HPCC-18142 Unbound DESDL based ESP services should report unbound state

### PR DESCRIPTION
**Pull request by mayx**
- Add a method to base and DESDL binding classes to indicate
  if the binding is dynamic binding
- Set the isDynamicBinding attribute in navigation data retrieved
- Show "Unbound DESDL Service" when there are no methods and the binding
  is dynamic.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [x] The component(s) render as expected

### CODE REVIEW

**Comment by mayx** [comment]
screen shot 2017-08-17 at 1 14 24 pm

**Comment by mayx** [comment]
@afishbeck please review. Please check the attached image above for the look of the text.

**Comment by afishbeck** [comment]
@mayx The place where dynamicesdl2 shows up in the image has always seemed wrong to me.  That is where the first service name is supposed to go if there is one.  That name is the ESP binding name I think.

Can we get rid of that?  And maybe put the informational text right there?  "[No Services]" or if it fits "[No Service Bindings]".

**Comment by mayx** [comment]
@afishbeck Sorry didn't see your comments earlier. Actually "dynamicesdl2" here is the service name, not the binding name. I was using the same name for the esp service and binding. After changing the name of the service in environment.xml, now it looks as the image attached.

**Comment by mayx** [comment]
screen shot 2017-08-30 at 10 48 46 am

**Review by afishbeck — APPROVED** [review]
@mayx looks fine

File: esp/services/esdl_svc_engine/esdl_binding.hpp [hunk 1]
```diff
@@ -375,6 +375,7 @@ class EsdlBindingImpl : public CHttpSoapBinding
 
     bool usesESDLDefinition(const char * name, int version);
     bool usesESDLDefinition(const char * id);
+    virtual bool isDynamicBinding() { return true; }
```
Comment (richardkchapman) [inline esp/services/esdl_svc_engine/esdl_binding.hpp]: const and override modifiers would seem appropriate here

File: esp/platform/espbinding.hpp [hunk 2]
```diff
@@ -106,7 +106,7 @@ class esp_http_decl CEspBinding : public CInterface,
     virtual bool isValidServiceName(IEspContext & context, const char * name){return false;}
     virtual bool qualifyServiceName(IEspContext & context, const char * servname, const char * methname, StringBuffer & servQName, StringBuffer * methQName){return false;}
     virtual IRpcRequestBinding *createReqBinding(IEspContext &context, IHttpMessage *request, const char *service, const char *method){return NULL;}
-
+    virtual bool isDynamicBinding() { return false; }
```
Comment (richardkchapman) [inline esp/platform/espbinding.hpp]: const would seem appropriate here

**Comment by richardkchapman** [comment]
@mayx a couple of minor comments

**Pull request merged**